pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,572 @@
+        MPI_Op op = getMPIOpType(op_str);
+        PL_MPI_IS_SUCCESS(MPI_Allreduce(&sendBuf, &recvBuf, 1, datatype, op,
+                                        this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Allreduce wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer.
+     * @param op_str String of MPI_Op.
+     *
+     * @return recvBuf Receive buffer.
+     */
+    template <typename T>
+    auto allreduce(T &sendBuf, const std::string &op_str) -> T {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Op op = getMPIOpType(op_str);
+        T recvBuf;
+        PL_MPI_IS_SUCCESS(MPI_Allreduce(&sendBuf, &recvBuf, 1, datatype, op,
+                                        this->getComm()));
+        return recvBuf;
+    }
+
+    /**
+     * @brief MPI_Allreduce wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer vector.
+     * @param recvBuf Receive buffer vector.
+     * @param op_str String of MPI_Op.
+     */
+    template <typename T>
+    void Allreduce(std::vector<T> &sendBuf, std::vector<T> &recvBuf,
+                   const std::string &op_str) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Op op = getMPIOpType(op_str);
+        PL_MPI_IS_SUCCESS(MPI_Allreduce(sendBuf.data(), recvBuf.data(),
+                                        sendBuf.size(), datatype, op,
+                                        this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Allreduce wrapper.
+     *
+     * @tparam T C++ data type.
+     *
+     * @param sendBuf Send buffer vector.
+     * @param op_str String of MPI_Op.
+     *
+     * @return recvBuf Receive buffer.
+     */
+    template <typename T>
+    auto allreduce(std::vector<T> &sendBuf, const std::string &op_str)
+        -> std::vector<T> {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        MPI_Op op = getMPIOpType(op_str);
+        std::vector<T> recvBuf(sendBuf.size());
+        PL_MPI_IS_SUCCESS(MPI_Allreduce(sendBuf.data(), recvBuf.data(), 1,
+                                        datatype, op, this->getComm()));
+        return recvBuf;
+    }
+
+    /**
+     * @brief MPI_Barrier wrapper.
+     */
+    void Barrier() { PL_MPI_IS_SUCCESS(MPI_Barrier(this->getComm())); }
+
+    /**
+     * @brief MPI_Bcast wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer.
+     * @param root Rank of broadcast root.
+     */
+    template <typename T> void Bcast(T &sendBuf, int root) {
+        MPI_Datatype datatype = getMPIDatatype<T>();
+        PL_MPI_IS_SUCCESS(
+            MPI_Bcast(&sendBuf, 1, datatype, root, this->getComm()));
+    }
+
+    /**
+     * @brief MPI_Bcast wrapper.
+     *
+     * @tparam T C++ data type.
+     * @param sendBuf Send buffer vector.
+     * @param root Rank of broadcast root.
+     */
+    template <typename T> void Bcast(std::vector<T> &sendBuf, int root) {
```

Comment:
Here `sendBuf` is only meaningful in the `root` process. We could broadcast the size and allocate correctly in all processes before broadcasting the buffer. Might be nice to have it as a `std::vector<T>`-returning function.

Reply:
as discussed, we'll keep current implementation.